xinference/model/llm/ggml/tests/test_ctransformers.py

```diff
@@ -0,0 +1,246 @@
+)
+def test_ctransformer_init(model_spec, model_family):
+    from ctransformers import AutoConfig
+
+    quantization = "q4_0"
+    uid = "".join(random.choice(string.digits) for i in range(15))
+    path = "".join(
+        random.choice(string.ascii_letters + string.punctuation) for i in range(100)
+    )
+    model = MockCtransformersModel(
+        model_uid=uid,
+        model_family=model_family,
+        model_spec=model_spec,
+        quantization=quantization,
+        model_path=path,
+        ctransformers_Model_Config=None,
+    )
+
+    assert model.model_uid == uid
+    assert model.quantization == quantization
+    assert model.model_path == path
+    assert model._ctransformer_model_config is not None
+    assert isinstance(model._ctransformer_model_config, AutoConfig)
+
+    assert isinstance(model.model_spec, GgmlLLMSpecV1)
+    assert isinstance(model.model_family, LLMFamilyV1)
+    assert isinstance(model.model_family.model_specs[0], GgmlLLMSpecV1)
+
+    assert (
+        model.model_family.model_specs[0].model_format == model.model_spec.model_format
+    )
+    assert model.model_family.model_specs[0].model_format == model_spec.model_format
+    assert (
+        model.model_family.model_specs[0].model_size_in_billions
+        == model.model_spec.model_size_in_billions
+    )
+    assert (
+        model.model_family.model_specs[0].model_size_in_billions
+        == model_spec.model_size_in_billions
+    )
+    assert (
+        model.model_family.model_specs[0].quantizations
+        == model.model_spec.quantizations
+    )
+    assert model.model_family.model_specs[0].quantizations == model_spec.quantizations
+    assert model.model_family.model_specs[0].model_id == model.model_spec.model_id
+    assert model.model_family.model_specs[0].model_id == model_spec.model_id
+    assert (
+        model.model_family.model_specs[0].model_file_name_template
+        == model.model_spec.model_file_name_template
+    )
+    assert (
+        model.model_family.model_specs[0].model_file_name_template
+        == model_spec.model_file_name_template
+    )
+    assert model._llm is None
+
+
+@pytest.mark.parametrize(
+    "model_spec, model_family", [(mock_model_spec, mock_model_family)]
+)
+def test_model_generate(model_spec, model_family):
+    quantization = "q4_0"
+    uid = "".join(random.choice(string.digits) for i in range(100))
+    path = "".join(
+        random.choice(string.ascii_letters + string.punctuation) for i in range(100)
+    )
+    model = MockCtransformersModel(
+        model_uid=uid,
+        model_family=model_family,
+        model_spec=model_spec,
+        quantization=quantization,
+        model_path=path,
+        ctransformers_Model_Config=None,
+    )
+
+    assert model._llm is None
+
+    model.load()
+    assert isinstance(model._llm, MockPipeline)
+
+    # generate with stream
+    pattern = r"[0-4]"
+    for completion in model.generate_stream():
```

Comment:
It is meaningless to test against a mock method.

Reply:
updated with opt test